zerver/lib/actions.py

```diff
@@ -5282,6 +5294,32 @@ def do_change_user_setting(
             active_user_ids(user_profile.realm_id),
         )
 
+    if setting_name == "email_address_visibility":
+        person = dict(
+            user_id=user_profile.id,
+            email_address_visibility=user_profile.email_address_visibility,
+        )
+        send_event(
+            user_profile.realm,
+            dict(type="realm_user", op="update", person=person),
+            active_user_ids(user_profile.realm_id),
+        )
```

Comment:
We generally never do a `send_event` call before changing the actual values in the database; so we should reorder this block to follow that convention.

Reply:
We are sending event for each field after updating it. Do we want to first update `email` and then send events for updating `email_address_visibility`, `email` and `avatar_url` at last?

Reply:
All events should be only after the database update has been committed that it is reflecting to clients; it's I suppose OK to have update1-event1-update2-event2. and also we should use the `transaction.on_commit`  approach to make sure events are only sent if the transaction commits.  But the second part is not critical -- or at least a bug we have everywhere here.